### PR TITLE
Add default threading for windows

### LINKED ISSUE #471: Threading should have a default Windows implementation
In the file `library/threading.c`, a default implementation of the mutex is implemented for use on a Unix/POSIX platform that supports pthreads, but no other defaults are provided. A default Windows implementation would be a good idea, as it can be tested as part of the mbed TLS CI system, saves others reimplementing a common feature and would provide consistent support.

Suggested by Fredrik Axelsson, and open for community contribution.

### PR DESCRIPTION
**Pull request by mrsshr**
This PR fixes #471.

Uses critical section for synchronization.

Any ideas for spin count?

### CODE REVIEW

**Comment by RonEld** [comment]
Hi @mrsshr Thanks for the contribution!
Before we could start reviewing, our policy is to not accept contributions, without a Contributor’s Licence Agreement (CLA) signed or authorised by yourself or your employer.
If this is a personal contribution, the easiest way to do this is if you create an mbed account and [accept this click through agreement](https://developer.mbed.org/contributor_agreement/). Alternatively, you can find a [slightly different agreement to sign here](https://www.mbed.com/en/about-mbed/contributor-license-agreements/), which can be signed and returned to us, and is applicable if you don't want to create an mbed account or alternatively if this is a corporate contribution.
Thanks for your understanding and again, thanks for the contribution!

Please accept my apologies if I missed your signed agreement

**Comment by mrsshr** [comment]
Hi RonEld.

I am already agreed Contributor Agreement.

My profile is [here](https://developer.mbed.org/users/mrsshr/).

**Comment by RonEld** [comment]
Hi @mrsshr Thank you for your contribution!
The CI fails because at the following two scripts:
~~~
tests/scripts/check-generated-files.sh
tests/scripts/check-names.sh
~~~
In addition, the compilation fails  because:
~~~
error: "MBEDTLS_THREADING_WINDOWS defined, but not all prerequisites"
~~~
There are already too many configuration options in the *`config.h`* file. We don't tend to add new configurations, unless there is a strong justification for it. This new configuration is platform specific, which is also something we prefer to avoid. I thionk that perhaps you could use `MBEDTLS_THREADING_PTHREAD`, to define it as a platofrm threading implementation, (although it is not really pthread), and in your implementation, it will be within `#if defined(_WIN32)` check.
I don't think it's ideal, since it's not pthread, but it's better than introducing a new platform specific configuration option

**Comment by mrsshr** [comment]
i agree to your opinion
some code changed

and how to pass 'check-names.sh' test?
WIN32_LEAN_AND_MEAN for avoid socket header redefine
_WIN32_WINNT for critical section function [ref](https://msdn.microsoft.com/ko-kr/library/windows/desktop/ms683476(v=vs.85).aspx)
> To compile an application that uses this function, define _WIN32_WINNT as 0x0403 or later.

File: library/threading.c [hunk 1]
```diff
@@ -30,6 +30,9 @@
 #include "mbedtls/threading.h"
 
 #if defined(MBEDTLS_THREADING_PTHREAD)
```
Comment (sbutcher-arm) [inline library/threading.c]: Placing the code here, makes it dependent on `MBEDTLS_THREADING_PTHREAD` being defined which seems unnecessary.

If `MBEDTLS_THREADING_C` is defined and the target platform is Windows, I think that's sufficient.

File: library/threading.c [hunk 2]
```diff
@@ -79,6 +82,50 @@ int (*mbedtls_mutex_unlock)( mbedtls_threading_mutex_t * ) = threading_mutex_unl
  */
 #define MUTEX_INIT  = { PTHREAD_MUTEX_INITIALIZER, 1 }
 
+#else
```
Comment (sbutcher-arm) [inline library/threading.c]: I think it's a little confusing and unnecessary to have these functions defined in `threading.c`. I think I'd propose having them in `threading_windows.c` and rename `threading.c` to `threading_pthread.c` to make the code a bit clearer.

File: library/threading.c [hunk 3]
```diff
@@ -79,6 +82,50 @@ int (*mbedtls_mutex_unlock)( mbedtls_threading_mutex_t * ) = threading_mutex_unl
...
+    (void) DeleteCriticalSection( &mutex->cs );
+    mutex->is_valid = 0;
+}
+
+static int threading_mutex_lock_windows( mbedtls_threading_mutex_t *mutex )
+{
+    if( mutex == NULL || !mutex->is_valid )
+        return( MBEDTLS_ERR_THREADING_BAD_INPUT_DATA );
+
+    (void) EnterCriticalSection( &mutex->cs );
```
Comment (sbutcher-arm) [inline library/threading.c]: Why a cast to `void` here? The function returns no result.

File: library/threading.c [hunk 4]
```diff
@@ -79,6 +82,50 @@ int (*mbedtls_mutex_unlock)( mbedtls_threading_mutex_t * ) = threading_mutex_unl
...
+
+    mutex->is_valid = InitializeCriticalSectionAndSpinCount( &mutex->cs, 1024 ) != FALSE;
+}
+
+static void threading_mutex_free_windows( mbedtls_threading_mutex_t *mutex )
+{
+    if( mutex == NULL || !mutex->is_valid )
+        return;
+
+    (void) DeleteCriticalSection( &mutex->cs );
```
Comment (sbutcher-arm) [inline library/threading.c]: Why a cast to void here? The function returns no result.

File: library/threading.c [hunk 5]
```diff
@@ -79,6 +82,50 @@ int (*mbedtls_mutex_unlock)( mbedtls_threading_mutex_t * ) = threading_mutex_unl
...
+    (void) EnterCriticalSection( &mutex->cs );
+    return( 0 );
+}
+
+static int threading_mutex_unlock_windows( mbedtls_threading_mutex_t *mutex )
+{
+    if( mutex == NULL || !mutex->is_valid )
+        return( MBEDTLS_ERR_THREADING_BAD_INPUT_DATA );
+
+    (void) LeaveCriticalSection( &mutex->cs );
```
Comment (sbutcher-arm) [inline library/threading.c]: Why a cast to void here? The function returns no result.

File: library/threading.c [hunk 6]
```diff
@@ -79,6 +82,50 @@ int (*mbedtls_mutex_unlock)( mbedtls_threading_mutex_t * ) = threading_mutex_unl
...
 #define MUTEX_INIT  = { PTHREAD_MUTEX_INITIALIZER, 1 }
 
+#else
+
+static void threading_mutex_init_windows( mbedtls_threading_mutex_t *mutex )
+{
+    if( mutex == NULL || mutex->is_valid )
+        return;
+
+    mutex->is_valid = InitializeCriticalSectionAndSpinCount( &mutex->cs, 1024 ) != FALSE;
```
Comment (sbutcher-arm) [inline library/threading.c]: MSDN claims `InitializeCriticalSectionAndSpinCount()` always succeeds on Vista and beyond, which is what we support in the `development` branch.

That implies the test with `FALSE` is redundant.

File: include/mbedtls/threading.h [hunk 7]
```diff
@@ -40,14 +40,34 @@ extern "C" {
...
 #include <pthread.h>
 typedef struct
 {
     pthread_mutex_t mutex;
     char is_valid;
 } mbedtls_threading_mutex_t;
+
+#else
+
+#if !defined(_WIN32_WINNT)
```
Comment (sbutcher-arm) [inline include/mbedtls/threading.h]: Do we really need this check?

File: include/mbedtls/threading.h [hunk 8]
```diff
@@ -40,14 +40,34 @@ extern "C" {
...
+#define _WIN32_WINNT 0x0403
 #endif
 
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+typedef struct
```
Comment (sbutcher-arm) [inline include/mbedtls/threading.h]: This `struct` should be in the threading header surely?

**Comment by mrsshr** [comment]
Hello, @SergeySeroshtan

@sbutcher-arm mentions that mbedtls does not support Windows XP.

I think this PR is useless because using SRWLock(Vista and later) is better then using critical section.

**Comment by SergeySeroshtan** [comment]
Add PR #2306 that is based on the SRW Locks implementation.

**Pull request closed without merging**